### PR TITLE
Hotfix/reactor 427 - ensure that timestamp is properly transmitted to Core pipeline

### PR DESCRIPTION
**Pull request by killthrush**
### Description

CZ-NRF requires a (poorly-named) `timestamp` field to be passed from Reactor to Core among other things.  The latest reactor architecture stopped transmitting it, leading to resource data being sent that was not entirely valid.  Fix this by sending the correct data.

- [ ] Existing GitHub Issue? Please provide #: 

### Acceptance checklist

- [ ] Feedback addressed
- [ ] Tests added or modified appropriately
- [ ] All tests passed

### CODE REVIEW

File: lambda_handlers.py [hunk 1]
```diff
@@ -133,13 +133,15 @@ def accounts__id__resources__post(event, context, namespace, configuration, clie
     except KeyError:
         return generate_response(400, json.dumps({'error': 'Missing account ID.'}))
 
-    result = run_probe_single_account(account_id, namespace, configuration)
+    reactor_account_id = get_account_id(context)
+    result = run_probe_single_account(account_id, namespace, configuration, reactor_account_id)
```
Comment (killthrush) [inline lambda_handlers.py]: While fixing this, found another bug with `reactor_account_id` - it was missing in a few places that needed them (account_context, mainly).  Refactored all the call sites to require it and updated tests.

File: reactor/features/accounts/account_context.py [hunk 2]
```diff
@@ -22,24 +22,26 @@
 logger = lambda_tools.setup_logging('reactor')
 
 
-def build_cloud_account_context(namespace, configuration, cloud_account_id=None, cz_account_id=None,
-                                reactor_account_id=None):
+def build_cloud_account_context(namespace, configuration, reactor_account_id,
+                                cloud_account_id=None, cz_account_id=None):
```
Comment (killthrush) [inline reactor/features/accounts/account_context.py]: This is no longer optional - reading from account settings is not reliable so support was dropped.

File: reactor/features/accounts/verify.py [hunk 3]
```diff
@@ -181,7 +181,10 @@ def _do_verification(details):
             # These _must_ work if anything is to work
             verify_responses['target_account'] = list(aws_tools.verify_account_access(aws))
             verify_responses['cloudtrail_account'] = list(aws_tools.verify_cloudtrail_access(details))
-            verify_responses['cloudtrail_bucket'] = list(aws_tools.verify_cloudtrail_bucket_access(aws, details))
+
+            # HACK ALERT! work around glacier problem until REACTOR-424 is resolved
+            verify_responses['cloudtrail_bucket'] = [ResourceAccess.OK, 'Because I said so']
+            # verify_responses['cloudtrail_bucket'] = list(aws_tools.verify_cloudtrail_bucket_access(aws, details))
```
Comment (killthrush) [inline reactor/features/accounts/verify.py]: Here's Erik's temporary fix for the bucket verify issue

File: reactor/features/resources/enrich.py [hunk 4]
```diff
@@ -193,8 +193,11 @@ def enrich_resource_observations(resource_observations, namespace, reactor_accou
                 resource_id=unenriched_resource['res_id'],
                 service_type=service_type,
                 resource_type=resource_type,
+                cloud_local_id=unenriched_resource['cloud_local_id'],
                 cloud_region=unenriched_resource['region'],
                 cloud_account_id=unenriched_resource['cloud_account_id'],
+                last_observed=unenriched_resource['last_observed'],
+                observation_count=unenriched_resource['observation_count'],
```
Comment (killthrush) [inline reactor/features/resources/enrich.py]: Ensure that by default we include observation fields and cloud local id.  The former were missing from coresync, and the `last_observed` date in particular is what we are going to send over with the `timestamp` label

File: reactor/features/resources/enrich.py [hunk 5]
```diff
@@ -273,12 +276,9 @@ def _run_enrichment_plugin(account_context, enrich_plugin, resource_stream_event
...
         logger.error(f'Parsed CZRN: {czrn.extract_all_components(resource_id)}')
-        enriched_resources = [{'res_id': resource_id,
-                               'environment_id': resource_stream_event['environment_id'],
-                               'data': data,
-                               'disposition': resource.DISPOSITION_ERROR,
-                               'resource_state': None,
-                               'cz_metadata': None}]
+        enriched_resources.append(resource.create_from_stream_event(resource_stream_event,
+                                                                    resource.DISPOSITION_NO_DATA,
+                                                                    data))
```
Comment (killthrush) [inline reactor/features/resources/enrich.py]: fix this - while testing noticed that the error handler was generating bad records

File: reactor/features/resources/reconcile.py [hunk 6]
```diff
@@ -133,6 +133,7 @@ def _normalize_incoming_resource(enriched_resource):
     new_resource['cz_account_id'] = enriched_resource['environment_id']
     new_resource['resource_id'] = enriched_resource['res_id']
     new_resource['count'] = enriched_resource['enrich_count']
+    new_resource['timestamp'] = enriched_resource.get('last_observed')
```
Comment (killthrush) [inline reactor/features/resources/reconcile.py]: just added this to make sure the output is compliant with CZ-NRF.  This is going to all be revised in a later ticket - the names of the fields need an overhaul.

File: reactor/features/accounts/account_context.py (hunk 2 above)
Comment (tiyberius) [reply reactor/features/accounts/account_context.py]: 👍

File: lambda_handlers.py (hunk 1 above)
Comment (imcgaunn) [reply lambda_handlers.py]: you were forced to wade directly into the weeds for this one. thank you for doing the cloud lord's work

File: reactor/features/resources/resource.py [hunk 7]
```diff
@@ -127,18 +127,21 @@ def create_default_from_stream_event(resource_stream_event):
...
+                                   cloud_account_id=resource_stream_event['cloud_account_id'],
+                                   environment_id=resource_stream_event['environment_id'],
+                                   cloud_global_id=resource_stream_event.get('cloud_global_id'),
+                                   timestamp=resource_stream_event.get('timestamp'),
+                                   res_id=resource_stream_event['res_id'],
+                                   raw_data=data,
+                                   disposition=disposition)
+    resource['last_observed'] = resource_stream_event['last_observed']
+    resource['observation_count'] = resource_stream_event['observation_count']
+    return resource
```
Comment (killthrush) [inline reactor/features/resources/resource.py]: This needed revision as well, as enrich plugins need to ensure that necessary values are carried forward

File: test/unit/features/accounts/test_account_context.py [hunk 8]
```diff
@@ -45,7 +45,8 @@ def context(mocker):
 
 @pytest.mark.unit
 def test_load_context_for_existing_cloud_account_id(context):
-    loaded_context = account_context.build_cloud_account_context('test-ns', {}, cloud_account_id='123456789012')
+    loaded_context = account_context.build_cloud_account_context('test-ns', {}, '098765432109',
+                                                                 cloud_account_id='123456789012')
```
Comment (killthrush) [inline test/unit/features/accounts/test_account_context.py]: update tests for reactor account ID fix

File: reactor/features/accounts/account_resources.py [hunk 9]
```diff
@@ -56,17 +57,21 @@ def run_probe_single_account(cloud_account_id, namespace, configuration):
...
     try:
-        account_context = build_cloud_account_context(namespace, configuration, cloud_account_id=cloud_account_id)
+        account_context = build_cloud_account_context(namespace,
+                                                      configuration,
+                                                      reactor_account_id,
+                                                      cloud_account_id=cloud_account_id)
 
         if not account_context:
-            return {'error': 'Missing environment', 'count': 0}
+            return {'error': 'Missing environment', 'observation_count': 0}
```
Comment (imcgaunn) [inline reactor/features/accounts/account_resources.py]: I support this name change

File: lambda_handlers.py (hunk 1 above)
Comment (killthrush) [reply lambda_handlers.py]: yep, it's all good!  These things all needed to be resolved anyway

File: reactor/features/resources/reconcile.py (hunk 6 above)
Comment (imcgaunn) [inline reactor/features/resources/reconcile.py]: :1st_place_medal:

File: test/unit/features/accounts/test_account_context.py [hunk 10]
```diff
@@ -63,7 +64,8 @@ def test_load_context_for_existing_cloud_account_id(context):
 
 @pytest.mark.unit
 def test_load_context_for_existing_cz_account_id(context):
-    loaded_context = account_context.build_cloud_account_context('test-ns', {}, cz_account_id='my_env_id')
+    loaded_context = account_context.build_cloud_account_context('test-ns', {}, '098765432109',
+                                                                 cz_account_id='my_env_id')
```
Comment (imcgaunn) [inline test/unit/features/accounts/test_account_context.py]: this one looks different than the other instances of `cz_account_id` in this file. Not sure if that matters.

File: reactor/features/accounts/verify.py (hunk 3 above)
Comment (imcgaunn) [reply reactor/features/accounts/verify.py]: should we create a ticket to remove this hack? Do we have one already?

Comment (Tankanow) [reply reactor/features/accounts/verify.py]: Good comment

Comment (killthrush) [reply reactor/features/accounts/verify.py]: Reactor-424 was the other fix I was working on, which will address this

**Pull request merged**